src/storage/secondary/rowset/rowset_iterator.rs

```diff
@@ -126,189 +125,157 @@ impl RowSetIterator {
         // TODO: parallel fetch
         // TODO: align unmatched rows
 
-        for id in 0..filter_column.len() {
-            if filter_column[id] {
-                if let Some((row_id, array)) = self.column_iterators[id]
-                    .as_mut()
-                    .unwrap()
-                    .next_batch(Some(fetch_size), None)
-                    .await?
-                {
-                    if let Some(x) = common_chunk_range {
-                        if x != (row_id, array.len()) {
-                            panic!("unmatched rowid from column iterator");
+        // The visibility_map would probably be changed twice during scan process:
+        //   1) if this rowset has delete vectors, that would be applied on it
+        //   2) if filter needed, the filtered rows would be absent from the map
+        // Otherwise, it retains `None` to indicate that at the current stage,
+        // we would scan all rows and won't skip any block.
+        let mut visibility_map = None;
+
+        // Generate the initial visibility_map from delete vectors, so
+        // that we can avoid unnecessary scan on filter column at next
+        if !self.dvs.is_empty() {
+            // Get the start row id first
+            let start_row_id = {
+                let mut row_id = 0;
+                for (id, column_ref) in self.column_refs.iter().enumerate() {
+                    match column_ref {
+                        StorageColumnRef::Idx(_) => {
+                            row_id = self.column_iterators[id]
+                                .as_ref()
+                                .unwrap()
+                                .fetch_current_row_id();
+                            break;
                         }
+                        _ => continue,
                     }
-                    common_chunk_range = Some((row_id, array.len()));
-                    arrays.push(Some(array));
-                } else {
-                    arrays.push(None);
                 }
-            } else {
-                arrays.push(None);
+                row_id
+            };
+
+            // Initialize visibility map and apply delete vector to it
+            let mut visi = BitVec::new();
+            visi.resize(fetch_size, true);
+            for dv in &self.dvs {
```

Comment:
The `visi` can remain `None` if `self.dvs` is empty. We can do this optimization later.